apps/backend/src/factory/StfcDownloadService.ts

```diff
@@ -0,0 +1,149 @@
+import querystring from 'querystring';
+import { Readable } from 'stream';
+import { ReadableStream } from 'stream/web';
+
+import { logger } from '@user-office-software/duo-logger';
+import contentDisposition from 'content-disposition';
+import { Request, Response, NextFunction } from 'express';
+import { container } from 'tsyringe';
+
+import { Tokens } from '../config/Tokens';
+import { FileDataSource } from '../datasources/IFileDataSource';
+import { Role } from '../models/Role';
+import {
+  MetaBase,
+  DownloadType,
+  PDFType,
+  XLSXType,
+  ZIPType,
+  DownloadService,
+} from './DownloadService';
+import { ProposalPDFData } from './pdf/proposal';
+
+const ENDPOINT = process.env.USER_OFFICE_FACTORY_ENDPOINT;
+
+export class StfcDownloadService implements DownloadService {
+  async callFactoryService<TData, TMeta extends MetaBase>(
+    downloadType: DownloadType,
+    type: PDFType | XLSXType | ZIPType,
+    properties: { data: TData[]; meta: TMeta; userRole: Role },
+    req: Request,
+    res: Response,
+    next: NextFunction
+  ) {
+    if (!ENDPOINT) {
+      logger.logError(
+        'Could not start application: the `USER_OFFICE_FACTORY_ENDPOINT` environment variable is missing. Exiting.',
+        {}
+      );
+      process.exit(1);
+    }
+
+    let isPdfAvailable = false;
+    let proposalPdfData = null;
+    const data = properties?.data[0] as ProposalPDFData;
+
+    if (properties?.data?.length === 1) {
+      const fileDataSource = container.resolve<FileDataSource>(
+        Tokens.FileDataSource
+      );
+      proposalPdfData = await fileDataSource.getBlobdata(
+        `Migrated-${data.proposal.proposalId}.pdf`
+      );
+
+      if (proposalPdfData) {
+        isPdfAvailable = true;
+      }
+    }
+
+    if (isPdfAvailable && proposalPdfData) {
+      try {
+        const fileName = `${data.proposal.proposalId}_${
+          data.principalInvestigator.lastname
+        }_${data.proposal.created.getUTCFullYear()}.pdf`;
+
+        res.setHeader('Content-Disposition', contentDisposition(fileName));
+        res.setHeader('x-download-filename', querystring.escape(fileName));
+        res.setHeader('content-type', 'application/pdf');
+
+        proposalPdfData.pipe(res);
+        await new Promise((f) => setTimeout(f, 1000));
+        proposalPdfData.emit('end');
+      } catch (error) {
+        next({
+          error,
+          message: 'Could not download generated',
+        });
+      }
+    } else {
+      const controller = new AbortController();
```

Comment:
From here, it is identical to the code in the EssDownloadService. Can we create a separate function like fetchDataAndStreamResponse(req, res, next, downloadType, type, properties) or something similar , and use it both here and there?